File: examples/leetcode-out/hs/30/substring-with-concatenation-of-all-words.hs
module Main where

import Data.Maybe (fromMaybe)
import qualified Data.Map as Map


forLoop :: Int -> Int -> (Int -> Maybe a) -> Maybe a
forLoop start end f = go start
  where
    go i | i < end =
            case f i of
              Just v -> Just v
              Nothing -> go (i + 1)
         | otherwise = Nothing

avg :: Real a => [a] -> Double
avg xs | null xs = 0
      | otherwise = sum (map realToFrac xs) / fromIntegral (length xs)

_indexString :: String -> Int -> String
_indexString s i =
  let idx = if i < 0 then i + length s else i
  in if idx < 0 || idx >= length s
       then error "index out of range"
       else [s !! idx]

_input :: IO String
_input = getLine


findSubstring :: String -> [String] -> [Int]
findSubstring s words = fromMaybe ([]) $
    case if (length words == 0) then Just ([]) else Nothing of Just v -> Just v; Nothing -> (let wordLen = length (words !! 0) in (let wordCount = length words in (let totalLen = (wordLen * wordCount) in case if (length s < totalLen) then Just ([]) else Nothing of Just v -> Just v; Nothing -> (let freq = Map.fromList [] in case foldr (\w acc -> case if (w in freq) then (let freq = ((freq !! w) + 1) in Nothing) else (let freq = 1 in Nothing) of Just v -> Just v; Nothing -> acc) Nothing words of Just v -> Just v; Nothing -> (let result = [] in case forLoop 0 wordLen (\offset -> (let left = offset in (let count = 0 in (let seen = Map.fromList [] in (let j = offset in Nothing))))) of Just v -> Just v; Nothing -> Just (result))))))

main :: IO ()
main = do
    return ()

File: examples/leetcode-out/swift/30/substring-with-concatenation-of-all-words.swift
import Foundation

func _sliceString(_ s: String, _ i: Int, _ j: Int) -> String {
	var start = i
	var end = j
	let chars = Array(s)
	let n = chars.count
	if start < 0 { start += n }
	if end < 0 { end += n }
	if start < 0 { start = 0 }
	if end > n { end = n }
	if end < start { end = start }
	return String(chars[start..<end])
}

func findSubstring(_ s: String, _ words: [String]) -> [Int] {
	let s = s
	let words = words
	
	if words.count == 0 {
		return []
	}
	let wordLen = words[0].count
	let wordCount = words.count
	let totalLen = wordLen * wordCount
	if s.count < totalLen {
		return []
	}
	var freq: [String: Int] = [:]
	for w in words {
		if freq[w] != nil {
			freq[w] = freq[w]! + 1
		} else {
			freq[w] = 1
		}
	}
	var result: [Int] = []
	for offset in 0..<wordLen {
		var left = offset
		var count = 0
		var seen: [String: Int] = [:]
		var j = offset
		while j + wordLen <= s.count {
			let word = _sliceString(s, j, j + wordLen)
			j = j + wordLen
			if freq[word] != nil {
				if seen[word] != nil {
					seen[word] = seen[word]! + 1
				} else {
					seen[word] = 1
				}
				count = count + 1
				while seen[word]! > freq[word]! {
					let lw = _sliceString(s, left, left + wordLen)
					seen[lw] = seen[lw]! - 1
					left = left + wordLen
					count = count - 1
				}
				if count == wordCount {
					result = result + [left]
					let lw = _sliceString(s, left, left + wordLen)
					seen[lw] = seen[lw]! - 1
					left = left + wordLen
					count = count - 1
				}
			} else {
				seen = [:]
				count = 0
				left = j
			}
		}
	}
	return result
}

func main() {
}
main()
